designs/2019-typescript-eslintrc-support/README.md

```diff
@@ -0,0 +1,351 @@
+- Start Date: 2019-11-18
+- RFC PR: https://github.com/eslint/rfcs/pull/50
+- Authors: Gareth Jones ([@g-rath](https://github.com/g-rath))
+
+# First-Class Support For `.eslintrc.ts`
+
+## Summary
+
+Support the ability to consume `.eslintrc.ts` configuration files natively.
+
+## Motivation
+
+Currently configuration settings are validated by JSON Schema, which while
+powerful does not have ubiquitous & high quality support across editors &
+tooling.
+
+It also tends to generate somewhat cryptic messages, especially when using
+complex elements such as combining `oneOf`, `anyOf`, and other such schemas.
+
+Meanwhile, TypeScript is a powerful & popular superset of Javascript that has
+high quality support as a first class language in the majority of editors, and
+is quickly gaining ubiquitous support in general.
+
+By allowing first-class support for `.eslintrc.ts`, the amount of configuration
+problems & general confusion over such experienced by users can be reduced,
+since it would enable us to provide tooling to type configuration files and so
+allow TypeScript to catch ESLint configuration errors at compile time, rather
+than run time.
+
+While we can already hack ways to use `.eslintrc.ts`, none of them lend
+themselves to ease-of-use in the same manner that ESLint encourages, and so
+first-class support for `.eslintrc.ts` will go long way towards encouraging
+community driven typing of ESLint plugins.
+
+Finally, ESLint is being adopted as the linter for TypeScript, meaning
+first-class support for `.eslintrc.ts` will help enable a positive feedback loop
+to help improve the two tools.
+
+There are some potential UX benefits for users that use & understand TypeScript,
+in the form
+
+## Detailed Design
+
+- Adds support for `.eslintrc.ts`
+- Requires uses to install `ts-node` manually relative to the `node_modules` of
+  the configuration file, per standard Node
+  [module resolution logic](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders)
+  - Failure to do so with a `.eslintrc.ts` results in ESLint throwing a fatal
+    error
+  - The error message should include understandable plain text instructions on
+    how to fix the error.
+- ESLint does not validate the configuration object for now.
+- `.eslintrc.ts` is second highest priority, after `.eslintrc.js`
+  - This matches the priority of the other supported types, and generally where
+    TypeScript files it for other tools i.e webpack
+
```

Comment:
I wonder which `tsconfig.json` would be used with `ts-node`.

Reply:
When no tsconfig is present, it just uses some sensible default config.
Otherwise it will attempt to find one. IIRC it does a similar thing to node_modules resolution - traverse up the tree until it finds a parent dir with a tsconfig.

Reply:
`ts-node` will look first for the `TS_NODE_PROJECT` env variable, which is one of the nice things to all this - the env variables can be set in the standard fashion before running ESLint.

`ts-node` supports all of its flags & options as [env variables](https://github.com/TypeStrong/ts-node#cli-and-programmatic-options). This will be further improved when/if the ability to provide settings via the (`tsconfig.json` is landed)[https://github.com/TypeStrong/ts-node/issues/4], since then you would only need to provide the `TS_NODE_PROJECTS` variable at most.

If that isn't set, it will look for the `tsconfig` closest to it's require.

Reply:
I had some issues in the past with `ts-node` when the project is using project references. I always had to call `require('ts-node').register()` manually so that I can specify to correct `tsconfig.json` path.

Reply:
I've not worked with project references a lot, but ultimately `ts-node` only compiles what is needs, so the `tsconfig.json` requirement can be a lot looser than for a whole project.

In addition, `ts-node` overrides the majority of programmatic configuration options that cause issue, so unless you're doing some wild conditional logic or importing into your `.eslintrc.ts`, this shouldn't be a problem.

Personally, I don't really expect there to be a big reason for such logic, unlike in configurations for things such as webpack & babel, but I've not been a big user of `.eslintrc.js`.

If you know of specific examples of configuration files that have imports or actual conditional logic, I'm happy to field-test them against my patch :)

Reply:
`ts-node` just merged in the [ability to configure `ts-node` via `tsconfig.json`](https://github.com/TypeStrong/ts-node/issues/4), which I  believe adequately covers this.

I'll probably update the RFC with a note at some point, but for now I'm going to resolve this - happy to unresolve it if you want to discuss further :)